### PR TITLE
feat(sdk-js): return 4xx api errors

### PR DESCRIPTION
**Pull request by eliashussary**
## Summary
This PR aims to address how `DirectusSDK` should handle errors return from the API.

### Current Behavior
Currently `axios` throws an exception when the request does not meet the following criteria;
```
validateStatus: function validateStatus(status) {
    return status >= 200 && status < 300;
  }
```
https://github.com/axios/axios/blob/59ab559386273a185be18857a12ab0305b753e50/lib/defaults.js#L79-L81

### Target Behavior
The `DirectusSDK` should  return API errors to the caller and not require the consumer to wrap in a `try/catch`. Exceptions are now reserved for critical errors such as `5xx` status codes

Related discussion: https://github.com/directus/directus/discussions/3482#discussioncomment-237732

### Changes to Note

- added additional default config to axios to modify `validateStatus`
- refactored `axios.create` to util `createAxiosInstance` 
  - allows us to import for testing handlers and keeps things a bit more dry
- added negative test case for `ItemHandler` 
  - not sure what our vision is here for negative tests, this is likely a good starting point 
  - if we decide later to move errors into a dedicated suite, that's cool too
- added `nock` dependency rather than stubbing `axios`
    - this offers us more complete testing, that will also test changes made to config such as `validateStatus`

Note: the test suite file is complaining about prettier errors. I did not run prettier and i pushed with `--no-verify` this way my PR is clean. we need to update that and cut another PR for it.

### CODE REVIEW

**Comment by rijkvanzanten** [comment]
I don't know if I agree with this 🤔 

I think it's actually quite nice for the SDK to throw an error on 4xx/5xx requests as it allows you to use try/catch in async/await blocks / promise rejects. The only downside is that it throws the full axios error, and not just the returned error by the API.

Can we meet in the middle and leave the SDK's api as follows:

```js
async function fetchArticles() {
  loading = true;

  try {
    articles = directus.items("articles").read("12435");
  } catch (err) {
    error = err;
  } finally {
    loading = false;
  }
}
```

but have it return `error.response.data.errors` by default instead of the full axios error object?

File: packages/sdk-js/src/index.ts [hunk 1]
```diff
@@ -19,13 +19,14 @@ import {
...
 class DirectusSDK {
 	axios: AxiosInstance;
 	private authOptions: AuthOptions;
 
 	constructor(url: string, options?: { auth: Partial<AuthOptions> }) {
-		this.axios = axios.create({
+		this.axios = createAxiosInstance({
 			baseURL: url,
 			withCredentials: true,
 		});
```
Comment (eliashussary) [inline packages/sdk-js/src/index.ts:32]: would you be open to optionally accepting either the `axios` config, or `axios`  instance?

this way a consumer has control over the http transport?

**Comment by eliashussary** [comment]
@rijkvanzanten i've updated based on your feedback, currently rebasing due to conflicts. 

please consider my comment to allow extending the axios config by the consumer

File: packages/sdk-js/src/index.ts (hunk 1 above)
Comment (rijkvanzanten) [reply packages/sdk-js/src/index.ts:32]: Yeah totally! We already expose the axios instance itself through `.axios`, so I see no problem with allowing to configure that root axios instance (or passing it) on initialization of the SDK 👍🏻

Comment (eliashussary) [reply packages/sdk-js/src/index.ts:32]: great, i'll add another commit to cover this

Comment (eliashussary) [reply packages/sdk-js/src/index.ts]: good to review again @rijkvanzanten

**Review by rijkvanzanten — APPROVED** [review]

**Pull request closed without merging**